File: euler579c/vertex.cpp
#include "stdafx.h"
#include "vertex.h"

using namespace std;

bool operator<(const vertex & lhs, const vertex & rhs)
{
	if (lhs.x != rhs.x) return lhs.x < rhs.x;
	if (lhs.y != rhs.y) return lhs.y < rhs.y;
	if (lhs.z != rhs.z) return lhs.z < rhs.z;

	return false;
}

bool operator !=(const vertex& lhs, const vertex& rhs)
{
	return lhs < rhs || rhs < lhs;
}


vertex operator-(const vertex& lhs, const vertex& rhs)
{
	return vertex(lhs.x - rhs.x, lhs.y - rhs.y, lhs.z - rhs.z);
}

vertex operator*(const vertex& v, long n)
{
	return vertex(v.x * n, v.y * n, v.z * n);
}

ostream& operator<<(ostream& os, const vertex & vertex)
{
	os << vertex.x << "," << vertex.y << "," << vertex.z;
	return os;
}

long long vertex::dist_squared_from(const vertex & other) const
{
	return
		(other.x - x)*(other.x - x) +
		(other.y - y)*(other.y - y) +
		(other.z - z)*(other.z - z);
}

File: euler579cs2/vertex.cs
using System;

namespace euler579cs2
{
    public class vertex : IEquatable<vertex>
    {
        public vertex(long x, long y, long z)
        {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public long x { get; }
        public long y { get; }
        public long z { get; }

        public bool Equals(vertex other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return x == other.x && y == other.y && z == other.z;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((vertex)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = x.GetHashCode();
                hashCode = (hashCode * 397) ^ y.GetHashCode();
                hashCode = (hashCode * 397) ^ z.GetHashCode();
                return hashCode;
            }
        }

        public override string ToString()
        {
            return $"{x},{y},{z}";
        }

        public static bool operator ==(vertex left, vertex right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(vertex left, vertex right)
        {
            return !Equals(left, right);
        }

        public static vertex operator +(vertex lhs, vertex rhs)
        {
            return new vertex(lhs.x + rhs.x, lhs.y + rhs.y, lhs.z + rhs.z);
        }


    }
}